porch/func/internal/podevaluator.go

```diff
@@ -160,59 +255,70 @@ func (pe *podEvaluator) EvaluateFunction(ctx context.Context, req *pb.EvaluateFu
 			},
 		},
 	}
+	klog.Infof("creating KRM function evaluator pod %q for %q", pod.Name, image)
 	err = pe.kubeClient.Create(ctx, pod)
 	if err != nil {
-		return nil, err
+		return client.ObjectKey{}, err
 	}
-	klog.Infof("created KRM function evaluator pod %v", pod.Name)
-
-	defer func() {
-		// TODO: keep the pod around for some time for potential reuse.
-		klog.Infof("deleting KRM function evaluator pod %v", pod.Name)
-		// We don't use ctx here, since we want the deletion to be completed even the context is canceled.
-		// If we use ctx here, the pod will be orphaned if we cancel the request (e.g. ctrl+C).
-		if err := pe.kubeClient.Delete(context.Background(), pod); err != nil {
-			klog.Warningf("failed to delete KRM function evaluator pod %v: %v", pod.Name, err)
-		}
-	}()
+	return client.ObjectKeyFromObject(pod), nil
+}
 
+// podIpIfRunningAndReady waits for the pod to be running and ready and returns the pod IP and a potential error.
+func podIpIfRunningAndReady(ctx context.Context, cl client.Client, podKey client.ObjectKey) (ip string, e error) {
+	var pod corev1.Pod
 	// Wait until the pod is Running
-	if err = wait.PollImmediate(100*time.Millisecond, 60*time.Second, func() (done bool, err error) {
-		err = pe.kubeClient.Get(ctx, client.ObjectKeyFromObject(pod), pod)
+	if e := wait.PollImmediate(100*time.Millisecond, 60*time.Second, func() (done bool, err error) {
+		err = cl.Get(ctx, podKey, &pod)
 		if err != nil {
 			return false, err
 		}
 		if pod.Status.Phase != "Running" {
 			return false, nil
 		}
 		for _, cond := range pod.Status.Conditions {
-			if cond.Type == corev1.PodReady {
-				return cond.Status == corev1.ConditionTrue, nil
+			if cond.Type == corev1.PodReady && cond.Status == corev1.ConditionTrue {
+				return true, nil
 			}
 		}
 		return false, nil
-	}); err != nil {
-		return nil, fmt.Errorf("error when waiting the pod to be ready: %w", err)
+	}); e != nil {
+		return "", fmt.Errorf("error when waiting the pod to be ready: %w", e)
 	}
+	return pod.Status.PodIP, nil
+}
 
-	podIP := pod.Status.PodIP
+func (pe *podEvaluator) EvaluateFunction(ctx context.Context, req *pb.EvaluateFunctionRequest) (*pb.EvaluateFunctionResponse, error) {
+	podKey, err := pe.retrieveOrCreatePod(ctx, req.Image)
+	if err != nil {
+		return nil, err
+	}
+	podIP, err := podIpIfRunningAndReady(ctx, pe.kubeClient, podKey)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I'd recommend making this the 'slow path' of the request, should we get a cache miss on "which gRPC client is handling this function?" If we don't find the client we can then spin up the pod and get its client.

There is also potential for 'thundering herd' problem - if multiple clients call the same (uncached) function we'll start spinning up multiple pods in parallel.

Reply:
We can do much of this as follow-ups

Reply:
It should have be resolved now. PTAL